Resolve SME feedback in best practices (#347)
osisoft/OCS-Docs: Content_Portal/How_To_Guides/Add-Org/roles.md

> # Roles

Reviewer: @osisoft/product-readiness-guild Please review

You must have the **Account Administrator** role to add and manage roles in a tenant.

## PI Core Counterpart

Roles in OCS are comparable to PI Identities in the PI Data Archive, or Identities in the PI AF Server. Throughout OCS, permissions are granted to Roles instead of directly to individual users or clients, in much the same way that Identities in PI Core are used to assign permissions for a set of users or clients.

## Roles Best Practices

1. Consider whether the read access granted by the Account Member role is acceptable for all users and clients in your tenant. If you plan to invite users from outside your organization, it may be important to limit the read access granted by the Account Member role.

1. Use caution when granting the Account Administrator role, ideally using a different role for users and clients that should not be able to manage permissions. Clients should not generally be granted the Account Administrator role.

> 1. If users from other organizations will be given a user account to log in to your tenant, consider creating a role for their specific organization so that their permissions can be managed explicitly.

Reviewer 2: What happens if you have deny and read from different group?

What happens if you don't have a permission specified?

Reviewer: 1 - Added a new best practice to describe Deny, check the new fourth item
2 - Added a note to this at line 15 where I describe adding a new role, let me know if you think it needs more

# Roles

## Definition

Roles are used to manage access to assets, resources, and services in OSIsoft Cloud Services (OCS). Roles are assigned to identities, which includes users, groups, and client-credentials clients.

There are five built-in roles, which cannot be removed from a tenant.

> - **Account Administrator** - Represents the Administrator of OCS, and granted full permissions throughout OCS by default.

Author: ```suggestion
- **Account Administrator** - OCS Administrator of OCS who is granted full permissions throughout OCS, by default.
```

> - **Account Contributor** - Granted read and write permissions throughout OCS by default.

Author: ```suggestion
- **Account Contributor** - Granted read and write permissions throughout OCS, by default.
```

> - **Account Data Steward** - Not granted any specific permissions by default.

Author: ```suggestion
- **Account Data Steward** - No specific permissions  are granted to this role, by default.
```

Reviewer: There is an extra whitespace between `permissions` and `are` in this suggestion.

Author: Done. Committed changes.

> - **Account Viewer** - Not granted any specific permissions by default.

Author: ```suggestion
- **Account Viewer** - No specific permissions  are granted to this role, by default.
```

Reviewer: There is an extra whitespace between `permissions` and `are` in this suggestion.

Author: Done. Committed changes.

> - **Account Member** - Automatically granted to any user or client in OCS, and granted read access throughout OCS by default.

Author: ```suggestion
- **Account Member** - By default, this role is assigned to all users or client in OCS.  Account members are granted read access throughout OCS, by default.
```

Reviewer: I would remove `By default` since this is not configurable. `This role is assigned to all users` would be fine.

Author: Done. Committed changes.

> In addition, custom roles can be added, which are not granted any specific permissions by default. Granting a user or client the new custom role will not have any effect on permissions until the role is explicitly allowed or denied access to resources in OCS.

Author: ```suggestion
In addition, you can create custom roles which are not granted any specific permissions, by default. Access to OCS resources depends on explicitly granting or denying access to a role. Therefore, simply assigning a custom role to a user or client does not determine access. This is defined when a role is explicitly allowed or denied access to OCS resources.
```

Reviewer: Where you edited to `explicitly granting or denying access` we might want to consider saying `explicitly allowing or denying access`. The UI describes the two columns as `Allow` and `Deny` so it might be good to be consistent with the verbiage here.

Author: Done. Committed change.

> For any resource in OCS, permissions are granted or denied to specific roles, rather than specific users or clients. These permissions are managed using the **Manage Permissions** dialog for the given resource. Role access to a resource an include **Read**, **Write**, **Delete**, and **Manage Permissions**. Each access type can be set to **Allow** or **Deny** for each individual role.

Author: ```suggestion
For any resource in OCS, permissions are granted or denied to specific roles, rather than to specific users or clients. These permissions are managed using the **Manage Permissions** dialog for the given resource. Each role can be granted or denied one or more of the following access types:  **Read**, **Write**, **Delete**, and **Manage Permissions**.

Reviewer: Similar, it might be better to say `allowed or denied` instead of `granted or denied` to use consistent wording.

Author: Done. Committed change.

You must have the **Account Administrator** role to add and manage roles in a tenant.

## PI Core counterpart

> Roles in OCS are comparable to PI Identities in the PI Data Archive, or Identities in the PI AF Server. Throughout OCS, permissions are granted to Roles instead of directly to individual users or clients, in much the same way that Identities in PI Core are used to assign permissions for a set of users or clients.

Author: ```suggestion
Roles in OCS are comparable to PI identities in PI Data Archive or identities in PI AF server. Throughout OCS, permissions are granted to roles instead of directly to individual users or clients. This is similar to how identities in PI Core are used to assign permissions for a set of users or clients.
```

## Roles best practices

> 1. Consider whether the read access granted by the Account Member role is acceptable for all users and clients in your tenant. If you plan to invite users from outside your organization, it may be important to limit the read access granted by the Account Member role.

Author: What do you mean by "acceptable"? Do you mean sufficient? 
How do you limit read access to the Account Member? Isn't this the minimum permissions that a user can have?

Reviewer: What I mean is that if a company invites an external user into their tenant, they may not want them to have read access to everything. The read access granted to account member can be removed throughout OCS, but it has to be done manually. You can't remove the actual account member role from the user but you can completely remove any access granted to the account member role.

> 1. Use caution when granting the Account Administrator role, ideally using a different role for users and clients that should not be able to manage permissions. Clients should not generally be granted the Account Administrator role.

Author: ```suggestion
1. Use caution when granting the Account Administrator role.  OSIsoft recommends that you assign a different role to users and clients who should not manage permissions. OSIsoft recommends that you avoid assigning the Account Administrator role to clients.
```
Should we explain why they should not grant clients the Account Administrator role?

Reviewer: This should probably read `to client-credentials clients` to be more specific. And yes, we should probably say why, in fact maybe we should break this second sentence into its own best practice.
`1. Ensure that client-credentials clients are granted only the minimum set of permissions that are required by the application that uses them. This helps to minimize potential damage in case a client secret is compromised or a problem is encountered with the application.`

Author: Done. Committed changes.

1. If users from other organizations will be given a user account to log in to your tenant, consider creating a role for their specific organization so that their permissions can be managed explicitly.

> 1. Use caution when denying permissions, as denying supersedes any allowed access to a role. In other words, if a user is allowed access via one role and denied access via another role, the user will not have access. For this reason, it is not allowed to deny permissions to the **Account Member** role, as this would deny the given permission to every user in the tenant.

Author: ```suggestion
1. Use caution when denying permissions because this supersedes any allowed access to a role. For example, if a user is allowed write access through one role but is denied write access through another role, the user will not have write access. Because all users and clients are assigned the Account Member role, you cannot deny permissions to the **Account Member** role. Doing so would deny the given permission to every user in the tenant.
```
I'm confused by this.

# Roles

## Definition

Roles are used to manage access to assets, resources, and services in OSIsoft Cloud Services (OCS). Roles are assigned to identities, which includes users, groups, and client-credentials clients.

> There are five built-in roles, which cannot be removed from a tenant.

Author: ```suggestion
There are five built-in roles which cannot be removed from a tenant.
```

- **Account Administrator** - OCS Administrator of OCS who is granted full permissions throughout OCS, by default.
- **Account Contributor** - Granted read and write permissions throughout OCS, by default.
- **Account Data Steward** - No specific permissions are granted to this role, by default.
- **Account Viewer** - No specific permissions are granted to this role, by default.

> - **Account Member** - This role is assigned to all users or client in OCS.  Account members are granted read access throughout OCS, by default.

Author: ```suggestion
- **Account Member** - This role is assigned to all users or clients in OCS.  Account members are granted read access throughout OCS, by default.
```

In addition, you can create custom roles which are not granted any specific permissions, by default. Access to OCS resources depends on explicitly allowing or denying access to a role. Therefore, simply assigning a custom role to a user or client does not determine access. This is defined when a role is explicitly allowed or denied access to OCS resources.

For any resource in OCS, permissions are allowed or denied for specific roles, rather than to specific users or clients. These permissions are managed using the **Manage Permissions** dialog for the given resource. Each role can be allowed or denied access to one or more of the following access types:  **Read**, **Write**, **Delete**, and **Manage Permissions**. 

You must have the **Account Administrator** role to add and manage roles in a tenant.

## PI Core counterpart

Roles in OCS are comparable to PI identities in PI Data Archive or identities in PI AF server. Throughout OCS, permissions are granted to roles instead of directly to individual users or clients. This is similar to how identities in PI Core are used to assign permissions for a set of users or clients.

## Roles best practices

> 1. Consider whether the read access granted by the Account Member role is acceptable for all users and clients in your tenant. If you plan to invite users from outside your organization, it may be important to limit the read access granted by the Account Member role.

Author: ```suggestion
1. Consider whether the read access granted by the Account Member role is acceptable for all users and clients in your tenant. Specifically, if you plan to invite users from outside your organization, you may want to limit their read access. One way to do this is to create a custom role for external users so that their permissions can be managed explicitly.
```
I'm combining one of the other best practices with this one because they seem to be related.

1. Use caution when granting the Account Administrator role.  OSIsoft recommends that you assign a different role to users and clients who should not manage permissions. OSIsoft recommends that you avoid assigning the Account Administrator role to client-credentials clients. 
1. Ensure that the roles assigned to client-credentials clients only grant the minimum set of permissions required by the application that uses these clients. This minimizes the potential damage in the event a client secret is compromised or a problem arises with the application.

> 1. If users from other organizations will be given a user account to log in to your tenant, consider creating a role for their specific organization so that their permissions can be managed explicitly.

Author: ```suggestion
1. If users from other organizations will be given a user account to log in to your tenant, consider creating a role for their specific organization so that their permissions can be managed explicitly.
```
I merged this with another best practice above, so I think we can remove this one.

Reviewer: OK. It won't let me commit this suggestion, not sure if you already deleted it or what. I'm OK with removing it if it's a duplicate now.

Reviewer: Got this committed directly instead of using the suggestion.
